docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+[evolving protobuf schemas](https://developers.google.com/protocol-buffers/docs/proto3#updating)
+without correct [unknown field filtering](./adr-020-protobuf-transaction-encoding.md#unknown-field-filtering)
+
+Note that all the following discussion assumes that the proto file versioning and state machine versioning of a module
+are distinct in that:
+* proto files are maintained in a non-breaking way (using something like [buf breaking](https://docs.buf.build/breaking/overview)
+  to ensure all changes are backwards compatible)
+* proto file versions get bumped much less frequently, i.e. we might maintain `cosmos.bank.v1` through many versions
+  of the bank module state machine
+* state machine breaking changes are more common and ideally this is what we'd want to semantically version with
+go modules, ex. `x/bank/v2`, `x/bank/v3`, etc.
+
+### Problem 1: Semantic Import Versioning Compatibility
+
+Consider we have a module `foo` which defines the following `MsgDoSomething` and that we've released its state
+machine in go module `example.com/foo`:
+
+```protobuf
+package foo.v1;
+
+message MsgDoSomething {
+  string sender = 1;
+  uint64 amount = 2;
+}
+
+service Msg {
+  DoSomething(MsgDoSomething) returns (MsgDoSomethingResponse);
+}
+```
+
+Now consider that we make a revision to this module and add a new `condition` field to `MsgDoSomething` and also
+add a new validation rule on `amount` requiring it to be non-zero, and that following go semantic versioning we
+release the next state machine version of `foo` as `example.com/foo/v2`.
+
+```protobuf
+// Revision 1
+package foo.v1;
+
+message MsgDoSomething {
+  string sender = 1;
+  
+  // amount must be a non-zero integer.
+  uint64 amount = 2;
+  
+  // condition is an optional condition on doing the thing.
+  //
+  // Since: Revision 1
+  Condition condition = 3;
+}
+```
+
+Approaching this naively, we would generate the protobuf types for the initial
+version of `foo` in `example.com/foo/types` and we would generate the protobuf
+types for the second version in `example.com/foo/v2/types`.
+
+Now let's say we have a module `bar` which talks to `foo` using this keeper
+interface which `foo` provides:
+```go
+type FooKeeper interface {
+	DoSomething(MsgDoSomething) error
+}
+```
+
+#### Scenario A: Backward Compatibility: Newer Foo, Older Bar
+
+Imagine we have a chain which uses both `foo` and `bar` and wants to upgrade to
+`foo/v2`, but the `bar` module has not upgraded to `foo/v2`.
+
+In this case, the chain will not be able to upgrade to `foo/v2` until `bar` 
+has upgraded its references to `example.com/foo/types.MsgDoSomething` to 
+`example.com/foo/v2/types.MsgDoSomething`.
+
+Even if `bar`'s usage of `MsgDoSomething` has not changed at all, the upgrade
+will be impossible without this change because `example.com/foo/types.MsgDoSomething`
+and `example.com/foo/v2/types.MsgDoSomething` are fundamentally different
+incompatible structs in the go type system.
```

Comment:
I think there are several fundamental misunderstandings on display here.

1. As described above, a protobuf package foo.v1 would always be implemented by the same application package
2. Protobuf definitions of the same version must be backwards compatible and therefore interoperable

As a consequence of (1) any chain that wants to include packages that use different but backwards-compatible versions of the same protobuf definition will encounter no issue, as the code they import will have the same identifier, and the compilation unit establishes a single well-defined source of truth for that identifier. There is no (non-pathological) situation in which code that presents/expects a Go foo/v1.MsgDoSomething will try to interact with code that expects/presents a Go foo/v2.MsgDoSomething.

As a consequence of (2) it doesn't matter which (minor/sub/patch) version of the protobuf package foo.v1 was chosen, because it will be the same choice for both foo and bar in a given compilation unit, and either choice is necessarily acceptable by any remote service that will process it.

Reply:
I addressed this above, but again (1) isn't true. It's not the same corresponding app code. State machine breaking changes are important and represent different app code for the same protobuf package.

As for (2), interoperability is not satisfied by backwards compatibility alone. Correct forward compatibility is also important and I've tried to address the issues with that in the ADR.

Reply:
> State machine breaking changes are important and represent different app code for the same protobuf package.

[As described above](https://github.com/cosmos/cosmos-sdk/pull/11802#discussion_r1029928828) it's not valid for a single protobuf service definition — or any other uniquely identifiable IDL endpoint — to have multiple incompatible implementations that are accessible at the same time.

Reply:
Having different distinct state machine exposing a compatible client API at different minor versions is however totally valid and it's how way cosmos works from a client perspective today

Reply:
If those distinct state machines produce equivalent outcomes for the same input, sure. But if that's not the case, then it's definitely not valid, regardless of precedent 😅